Add support for subscriptions admin payment method changes

luizreis opened the pull request:
Fixes #700 

#### Changes proposed in this Pull Request

* Add support for subscription_payment_method_change_admin
* Display the shopper's tokens in a select element when editing a subscription
* ~Display user saved payment methods at the user view and edit page~ (moved to #900 )

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

These flows need to be tested using the `subscription-edit-page.js`, which is applied when using Subscriptions `<= 3.0.7`, and using the new `payment_meta_input_` action, which needs a Subscriptions version based on subscriptions#3798 and its version manually set to `3.0.8`or higher.

The error flows are more to check the validation of the metadata, as they're unlikely to happen during real usage.

1. **Succesful update**
    1.1. On the subscription edit page, click on the edit button for its billing details
    1.2. Scroll down until "Saved payment method", assert that the dropdown contains all the saved cards for this customer
    1.3. Select another card and update the subscription
    1.4. Make a renewal and assert that the selected card was used
2. **No token selected**
    2.1. On the billing details edit page, use the JS console to set the select value to an invalid one 
`document.getElementById( '_payment_method_meta[woocommerce_payments][wc_order_tokens][token]' ).value = 0`
    2.2. Click on update and assert that the `"A customer saved payment method was not selected for this order"` error pops up
3. **Invalid token**
    3.1. On the billing details edit page, use the JS console to create a new option for another user token 
    ```
    var opt = document.createElement( 'option' );
    opt.value = 584954;
    document.getElementById( '_payment_method_meta[woocommerce_payments][wc_order_tokens][token]' ).appendChild( opt );
    ```
    3.2. Select the new option and click on update
    3.3. Assert that the `"The saved payment method selected is invalid or does not exist"` error pops up
4. **Token from another customer**
    4.1. On the billing details edit page, use the JS console to create a new option for another user token 
    ```
    var opt = document.createElement( 'option' );
    opt.value = 145; // Change this for another user token ID
    document.getElementById( '_payment_method_meta[woocommerce_payments][wc_order_tokens][token]' ).appendChild( opt );
    ```
    4.2. Select the new option and click on update
    4.3. Assert that the `"The saved payment method selected does not belong to this order's customer"` error pops up

-------------------

- [x] Added changelog entry (or does not apply)
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

<!--
Please read P7bbVw-3ww-p2 before opening the PR, assign the correct status to it, __and make sure that the related issue uses the Has PR status!__.
-->

luizreis committed:
Add support for subscription_payment_method_change_admin

This commit adds support for subscription_payment_method_change_admin by displaying the active
token ID under "Saved payment method ID" in the subscriptions payment meta and saves any changes
to the order payment tokens.

luizreis committed:
Use Stripe's payment_method_id instead of token

Use Stripe's payment method id instead of token in the subscriptions admin update page
to match the data that's going to be displayed in the shopper's saved card list.

luizreis committed:
Display token list when changing a subscription payment method

This commit removes the default input field with the payment method ID, for a select
element with all the user's tokens as options, making it easier for merchants to update
a subscription.

To do so, it uses a new script (subscription-edit-page.js) to make the DOM changes.
This script is localized with the subscriptions and tokens data and is compiled to
avoid issues with older browsers.

luizreis committed:
Revert "Use Stripe's payment_method_id instead of token"

This reverts commit deeb8ca7fb0afcc894fbdf3ff27273b890a186aa.

Now that a token list is displayed to the merchants, there's no need to use
the payment method ID instead of the token ID, as those values can be filled
automatically.

luizreis committed:
Add constants for subscriptions payment method meta table and key

luizreis committed:
Add changelog entry

luizreis commented:
I removed the `Needs Review` label while we investigate if we can add a filter to Subscriptions and discard the JS manipulation for newer Subscriptions versions.

luizreis committed:
Use new Subscription action hook to admin payment meta inputs

This commit depends on the hook added in WC Subs #3798, which can
be used to provide a different meta input, rendering the JS from
subscription-edit-page.js unecessary. To support older Subscriptions
versions, the JS method was kept, but the plan is to remove it in the
future, when more merchants have migrated to a newer Subs version.

DanReyLop commented on includes/compat/subscriptions/class-wc-payment-gateway-wcpay-subscriptions-compat.php:
```diff
@@ -111,16 +122,146 @@ public function scheduled_subscription_payment( $amount, $renewal_order ) {
+				self::PAYMENT_METHOD_META_TABLE,
+				self::PAYMENT_METHOD_META_KEY
+			),
+			function ( $input, $field_id, $field_value, $field_disabled ) use ( $subscription ) {
```
The filter should provide `$subscription` as a parameter instead.

luizreis replied:
I like this, thanks for the suggestion. Fixed in https://github.com/Automattic/woocommerce-payments/pull/892/commits/81fb6765eaa092812055effe11a6d286680ce1dd.

luizreis committed:
Use $this->id instead of WC_Payment_Gateway_WCPay::GATEWAY_ID

luizreis committed:
Use $token->get_display_name instead of assembling it from the data

luizreis committed:
Fix registering of WCPAY_SUBSCRIPTION_EDIT_PAGE script

luizreis commented:
I'm marking this as `Needs review` since it's probably in its final form. There are two `Tentative` commits that depend on `subscriptions#3798`, but it's likely that they won't need to be changed. Some of the reasoning behind that change can be found in p1599749421173100-slack-CGGCLBN58.

In case the subscriptions PR takes a while to get merged, we'd be able to split this PR in two and merge the part that uses the `subscription-edit-page.js` script.

luizreis commented:
New update: the Subscriptions PR went through some changes, but it's going to be shipped soon. I've applied the necessary changes here, removing the `Tentative` commits and rebased this on the latest master as some code from #893 could be applied here.

luizreis committed:
Add payment meta input registration tests

This commit adds some tests around the new Subscriptions payment meta
action hook and the workaround for older Subscriptions versions, that
don't have it.

Commit:
Merge branch 'master' into add/700-admin-subscriptions-payment-method-changes

DanReyLop commented on includes/compat/subscriptions/class-wc-payment-gateway-wcpay-subscriptions-compat.php:
```diff
@@ -46,6 +50,15 @@ public function __construct( ...$args ) {
+		add_filter( 'woocommerce_subscription_validate_payment_meta', [ $this, 'validate_subscription_payment_meta' ], 10, 3 );
+		add_action( 'wcs_save_other_payment_meta', [ $this, 'save_meta_in_order_tokens' ], 10, 4 );
+		// Enqueue JS hack when Subscriptions does not provide the meta input filter.
+		if ( version_compare( WC_Subscriptions::$version, '3.0.7', '<=' ) ) {
```
Let's add a `TODO` [in the main class](https://github.com/Automattic/woocommerce-payments/blob/a2217e35de3ea3cc43354fdde8a03f398e94bad0/includes/class-wc-payments.php#L108) to remove the JS hack when we bump the minimum required Subscriptions version to 3.0.8 or higher.

luizreis replied:
Good call. Added in https://github.com/Automattic/woocommerce-payments/pull/892/commits/5e3416bc51c67878dc0f0ce66e25d713ce070aa6.

DanReyLop commented on client/subscription-edit-page.js:
```diff
@@ -0,0 +1,28 @@
+	} );
+	paymentMethodSelect.id = paymentMethodInputId;
+	paymentMethodSelect.name = paymentMethodInputId;
+	paymentMethodSelect.value = paymentMethodInput.value;
```
`value` can be the empty string, so in that case it would be good to add an `<option value="">Choose one...</option>` element as the first option so it's clear that no payment method is selected.

luizreis replied:
Good idea! Added in https://github.com/Automattic/woocommerce-payments/pull/892/commits/da3a59f2832fb9967a9a770395d217f6e32602bd.

DanReyLop commented on includes/compat/subscriptions/class-wc-payment-gateway-wcpay-subscriptions-compat.php:
```diff
@@ -120,6 +133,132 @@ public function update_failing_payment_method( $subscription, $renewal_order ) {
+	 * @throws Exception When $payment_meta is not valid.
+	 */
+	public function validate_subscription_payment_meta( $payment_method_id, $payment_meta, $subscription ) {
+		if ( $this->id !== $payment_method_id ) {
```
This variable name (`$payment_method_id`) is confusing, since "payment method ID" means a different thing than "payment gateway ID" everywhere else in the code. Do you think it would be a good idea to change it, or let's keep the same name it has in the Subscriptions extension?

luizreis replied:
Yup! Changed it to `payment_gateway_id` in https://github.com/Automattic/woocommerce-payments/pull/892/commits/7d2bb5465d6b20f6ef7e5cef8cb0e7870f72332d.

DanReyLop commented on includes/compat/subscriptions/class-wc-payment-gateway-wcpay-subscriptions-compat.php:
```diff
@@ -162,4 +301,21 @@ public function maybe_render_subscription_payment_method( $payment_method_to_dis
+	public function render_custom_payment_meta_input( $subscription, $field_id, $field_value ) {
+		$tokens = $this->get_user_formatted_tokens_array( $subscription->get_user_id() );
+		echo '<select name="' . esc_attr( $field_id ) . '" id="' . esc_attr( $field_id ) . '">';
+		foreach ( $tokens as $token ) {
```
Same comment here about adding a `<option value="">Choose one...</option>` if `$field_value` is empty.

luizreis replied:
Good idea! Added in https://github.com/Automattic/woocommerce-payments/pull/892/commits/2683549de85bec95c1fe0b6440075f874ae22f1f.

DanReyLop reviewed (changes requested):
Looks great! I've added a few minor comments, I'll test the PR once they're addressed (or rejected).

luizreis committed:
Add TODO comment for removing the Subscriptions <= 3.0.7 JS hack

luizreis committed:
Rename $payment_method_id to $payment_gateway_id

This commit renames $payment_method_id to $payment_gateway_id
in validate_subscription_payment_meta, as the former can be
misleading since that's used to describe the payment method
used by the shopper when paying for an order.

luizreis committed:
Add default select option if token ID is not valid in subscriptions-edit-page.js

luizreis committed:
Add default select option if token ID is not valid for payment meta input

luizreis commented:
> Looks great! I've added a few minor comments, I'll test the PR once they're addressed (or rejected).

Thanks for the review, @DanReyLop! I've addressed your comments, thanks for suggesting them.

luizreis committed:
Merge 'master' into add/700-admin-subscriptions-payment-method-changes

DanReyLop reviewed (approved):
Works perfectly. I didn't check all the error cases (they're sufficiently covered by unit tests), but I checked:
- The `<select>` is displayed and works even when using Subscriptions 3.0.7.
- The `<select>` is displayed and works when using Subscriptions `trunk` and setting its version to `3.0.8`. In that case, the `subscriptions-edit-page.js` isn't downloaded at all.
- I had an old subscription that for some reason didn't have a token saved in the subscription meta, so I could test that the "Please choose a payment method..." option appears in the `<select>`.

:shipit: !
